src/modules/accounting/staking_vaults.py

```diff
@@ -0,0 +1,244 @@
+    fee: int
+    sharesMinted: int
+    leaf: str
+    proof: List[str]
+
+
+VaultToValidators = dict[ChecksumAddress, list[Validator]]
+
+
+class StakingVaults(Module):
+    w3: Web3
+    ipfs_client: MultiIPFSProvider
+    cl: ConsensusClient
+    vault_hub: VaultHubContract
+
+    def __init__(self, w3: Web3, cl: ConsensusClient, ipfs: MultiIPFSProvider, vault_hub: VaultHubContract) -> None:
+        super().__init__(w3)
+
+        self.w3 = w3
+        self.ipfs_client = ipfs
+        self.cl = cl
+        self.vault_hub = vault_hub
+
+    def _load_vault(self, address: ChecksumAddress) -> StakingVaultContract:
+        return cast(
+            StakingVaultContract,
+            self.w3.eth.contract(
+                address=address,
+                ContractFactoryClass=StakingVaultContract,
+                decode_tuples=True,
+            ),
+        )
+
+    def get_vaults(self, blockstamp: BlockStamp) -> VaultsMap:
+        vault_count = self.vault_hub.get_vaults_count(blockstamp)
+        pending_deposits = self.cl.get_pending_deposits(blockstamp)
+        deposit_map = dict[str, int]()
+
+        for deposit in pending_deposits:
+            if deposit.withdrawal_credentials not in deposit_map:
+                deposit_map[deposit.withdrawal_credentials] = 0
+
+            deposit_map[deposit.withdrawal_credentials] += Web3.to_wei(int(deposit.amount), 'gwei')
+
+        vaults = VaultsMap()
+        for vault_ind in range(vault_count):
+            vault_socket = self.vault_hub.vault_socket(vault_ind, blockstamp)
+
+            balance_wei = self.w3.eth.get_balance(
+                self.w3.to_checksum_address(vault_socket.vault), block_identifier=blockstamp.block_hash
+            )
+
+            vault = self._load_vault(vault_socket.vault)
+            vault_in_out_delta = vault.in_out_delta(blockstamp)
+            print(vault_in_out_delta)
+            vault_withdrawal_credentials = vault.withdrawal_credentials(blockstamp)
+            print(vault_withdrawal_credentials)
+
+            pending_deposit = 0
+            if vault_withdrawal_credentials in deposit_map:
+                pending_deposit = deposit_map[vault_withdrawal_credentials]
+
+            fee = 0
+            vaults[vault_socket.vault] = VaultData(
+                vault_ind,
+                balance_wei,
+                vault_in_out_delta,
+                vault_socket.shares_minted,
+                fee,
+                pending_deposit,
+                vault_socket.vault,
+                vault_withdrawal_credentials,
+                vault_socket,
+            )
+
+        return vaults
+
+    @staticmethod
+    def connect_vault_to_validators(validators: list[Validator], vault_addresses: VaultsMap) -> VaultToValidators:
+        wc_vault_map = dict[str, VaultData]()
+        for vault_pk in vault_addresses:
+            wc_vault_map[vault_addresses[vault_pk].withdrawal_credentials] = vault_addresses[vault_pk]
+
+        result = VaultToValidators()
```

Comment:
defaultdict will make a little bit cleaner here

```suggestion
    result: VaultToValidators = defaultdict(list)
    for validator in validators:
        wc = validator.validator.withdrawal_credentials
        if wc in wc_vault_map:
            result[wc_vault_map[wc].address].append(validator)
```

Reply:
fixed